### ISSUE TITLE
[Bug]: "Prop `href` did not match". when using tailwindcss

### ISSUE THREAD

**Issue by philldev** [remix-run/remix#896]
### Which Remix packages are impacted?

- [X] `remix` (Remix core)
- [ ] `create-remix`
- [ ] `@remix-run/architect`
- [ ] `@remix-run/cloudflare-workers`
- [ ] `@remix-run/dev`
- [ ] `@remix-run/express`
- [ ] `@remix-run/netlify`
- [ ] `@remix-run/node`
- [ ] `@remix-run/react`
- [X] `@remix-run/serve`
- [ ] `@remix-run/server-runtime`
- [X] `@remix-run/vercel`

### What version of Remix are you using?

1.0.6

### What version of Node are you using? Minimum supported version is 14.

16.13.1

### Steps to Reproduce

1. create remix app.
2. add tailwindcss from this guide https://remix.run/docs/en/v1/guides/styling#tailwind.
3. add tailwind.css to app/root.tsx.
```
import styles from './tailwind.css'

export function links() {
	return [{ rel: 'stylesheet', href: styles }]
}
```
4. start dev server

### Expected Behavior

No warning message appears on browser console

### Actual Behavior

This warning message appears on console
```
Warning: Prop `href` did not match. Server: "/build/_assets/tailwind-DGHXG3YF.css" Client: "/build/_assets/tailwind-JUD5S62U.css"
    at link
    at Links (http://localhost:3000/build/_shared/chunk-KOXENSZ3.js:3042:7)
    at head
    at html
    at Document (http://localhost:3000/build/root-WYXGDNDR.js:55:3)
    at App
    at RemixCatchBoundary (http://localhost:3000/build/_shared/chunk-KOXENSZ3.js:1476:10)
    at RemixErrorBoundary (http://localhost:3000/build/_shared/chunk-KOXENSZ3.js:1415:5)
    at RemixRoute (http://localhost:3000/build/_shared/chunk-KOXENSZ3.js:2905:3)
    at Routes (http://localhost:3000/build/_shared/chunk-KOXENSZ3.js:2888:7)
    at Router2 (http://localhost:3000/build/_shared/chunk-KOXENSZ3.js:457:21)
    at RemixCatchBoundary (http://localhost:3000/build/_shared/chunk-KOXENSZ3.js:1476:10)
    at RemixErrorBoundary (http://localhost:3000/build/_shared/chunk-KOXENSZ3.js:1415:5)
    at RemixEntry (http://localhost:3000/build/_shared/chunk-KOXENSZ3.js:2785:12)
    at RemixBrowser (http://localhost:3000/build/_shared/chunk-KOXENSZ3.js:3262:42)
```

**Comment by merelinguist** [comment]
A link to your code/a reproduction would be useful

**Comment by 0x219** [comment]
There is indeed this problem. Turning on disable cache can temporarily eliminate the warning.

**Comment by philldev** [comment]
> A link to your code/a reproduction would be useful

currently working on local right now.

**Comment by philldev** [comment]
> There is indeed this problem. Turning on disable cache can temporarily eliminate the warning.
@0x219 sure I'll try that

**Comment by nathanlindsay** [comment]
I don't know enough what's happening here to actually help fix this but the workaround I've been using is...

Update `package.json`:
```json
"dev:css": "tailwindcss -o ./public/tailwind.css --watch",
```

Then update `root.jsx`:
```javascript
export function links() {
	return [{ rel: "stylesheet", href: "/tailwind.css" }];
}
```

**Comment by yesmeck** [comment]
The strangest thing is that if you change the file name of `tailwind.css`, the error disappears.

```patch
-"dev:css": "tailwindcss -o ./app/tailwind.css --watch",
+"dev:css": "tailwindcss -o ./app/tailwindy.css --watch",
```

**Comment by Graham42** [comment]
Potentially related to https://github.com/remix-run/remix/issues/783

**Comment by AhmedBHameed** [comment]
I confirm this issue, however, it seems more like a cache issue only.
For anyone come across this issue, deleting `.cache` folder and re-run the server will fix the issue.

**Comment by wcly** [comment]
try move `tailwind.css` into `app/styles` by use command 
```
"dev:css": "tailwindcss -o ./app/styles/tailwind.css --watch"
```

**Comment by jesse996** [comment]
> The strangest thing is that if you change the file name of `tailwind.css`, the error disappears.
> 
> ```diff
> -"dev:css": "tailwindcss -o ./app/tailwind.css --watch",
> +"dev:css": "tailwindcss -o ./app/tailwindy.css --watch",
> ```

this works!

**Comment by arnaudambro** [comment]
It happened to me too, when I started creating a `styles` folder with other `.css` files inside.

> try move `tailwind.css` into `app/styles` by use command
> 
> ```
> "dev:css": "tailwindcss -o ./app/styles/tailwind.css --watch"
> ```

this workaround is correct for me !

**Comment by arnaudambro** [comment]
it's correct, although one more thing : it happened again to me even with the workaround.
deleting `.cache` didn't solve anything that time.
but when I hard refresh my page, everything went fine again : I guess it was a browser cache "issue" !

**Comment by kiliman** [comment]
I think I figured out why we're getting the error `Warning: Prop href did not match. Server: "/build/_assets/tailwind-UF2CZ54Y.css" Client: "/build/_assets/tailwind-GCIKYTPB.css`

It looks like when esbuild compiles the route file, it generates the hash based on the *contents of the original source*, **not** *the bundled output*. So if your root route imports the tailwind.css file from app, the source never changes. But the tailwind URL will change and will generate a bundle with the new file path, but with the SAME root hash! And since the /public/build folder is immutable, it will never download the new root file. :scream: The only way to work around this in dev, is to disable caching in your dev tools. But you can't rely on that on production.

To reproduce, have a clean /public/build folder. Start your app and let it build. Then rename /public/build to /public/build-old. Now add a new Tailwind class to your route and save. Remix will rebuild, but if you compare both root.js files (build and build-old), you'll see the filename stays the same even though the import for tailwind.css has changed.

```ts
// build-old/root-AURVXRN.js
// app/tailwind.css
var tailwind_default = "/build/_assets/tailwind-A6AN5OWU.css";

// build/root-AURVXRN.js. NOTE Same filename
// app/tailwind.css
var tailwind_default = "/build/_assets/tailwind-K3O5UDQR.css";
```
Not sure if this is an esbuild bug (it doesn't really say whether the hash is dependent on the source contents or the bundled contents).

**Comment by kiliman** [comment]
NOTE: I was able to reproduce this bug even without dynamic files like tailwind. A simple static CSS file will still result in the same error. So basically any import of static assets will fail because the route module hash will remain the same even if the imported filename changes.

**Comment by kiliman** [comment]
This is tracked and fixed in this esbuild issue https://github.com/evanw/esbuild/issues/1957

**Comment by gaarf** [comment]
seeing this issue a lot. hard refresh in dev fixes it.

**Comment by jan-cibulka** [comment]
Having this issue, without using tailwind. Sometimes, on first load of route, correct stylesheet is loaded, but when navigating back and forth, this issue appears and different version (built previously with different hash) of the same stylesheet is loaded.

**Comment by silvenon** [comment]
@gaarf for me it comes back after another edit. Hard refresh permanently fixes it for you?

**Comment by gaarf** [comment]
@silvenon after some edits yeah, the bug comes back eventually…. Hard
refresh fixes it until next time ;-) There’s no real impact that I can see
outside the console log.

**Comment by kiliman** [comment]
Any time you change your dependent file (.css), you'll get this error. Mainly because the esbuild version used by Remix isn't properly generating the filename hash for the importing file.

root-abc123.js imports style-def456.css

after changing css file, it creates a file named style-xyz789.css, which is imported in root.js correctly, but the filename is the same root-abc123.js (same filename, different content). Since these files are cached forever, your server side (which uses the new css) and your client (which is still using the old version... since it's been cached) is out of sync.

With that being said, the bug is fixed in latest esbuild, so hopefully it will be included in next remix release.

**Comment by adamlbunch** [comment]
> There is indeed this problem. Turning on disable cache can temporarily eliminate the warning.

How do you disable the cache?

**Comment by kiliman** [comment]
In Dev Tools, Network Tab
<img width="551" alt="image" src="https://user-images.githubusercontent.com/47168/152622584-ea0a6820-38cc-44e8-a132-78f8b3543516.png">

**Comment by divmgl** [comment]
> How do you turn on disabling the cache?

I would also add that you can use `Shift+Cmd+R` to do a single hard refresh.

> seeing this issue a lot. hard refresh in dev fixes it.

I can confirm that this is the workaround for now.

**Comment by kiliman** [comment]
This should be fixed in v1.2.2 (https://github.com/remix-run/remix/releases/tag/v1.2.2) since this release includes updated esbuild (50e428b65d0ec1ae72a19ea8a06bd8012fb01ce5) that fixes the underlying problem. @jacob-ebey

**Comment by silvenon** [comment]
I can confirm that this worked 🎉
